added a new function to rectify video playback (with ipad videos) in …

Lucinda-Williamson opened the pull request:
Could not properly process videos taken from ipads for Marathon data collection. Found a few issues like timestamps not being consistent and very high fps. This new function saves the original video files and then re-writes them to be compatible for viewing Jupyter Notebooks and in the dashboards.

Lucinda-Williamson committed:
added a new function to rectify video playback (with ipad videos) in notebooks/dashboards

Lucinda-Williamson commented on pose_pipeline/utils/video_format.py:
```diff
@@ -27,3 +28,57 @@ def insert_local_video(filename, video_start_time, local_path, video_project="TE
     Video().insert1(vid_struct, skip_duplicates=skip_duplicates)
+
+
+def make_browser_friendly(
```
It works in both. Right now written as a python function and previously had called it from a notebook. 
Example Usage (whether from notebook cell or .py script ran in terminal)
```
from pose_pipeline.utils.video_format import make_browser_friendly

output_path, backup_path = make_browser_friendly(
    filename="ma049_ipad_sagittal_right_treadmill_20250503.MP4",
    base_dir="/mnt/CottonLab/datasets/marathon_ipad/marathon_ipad_pre_training_2025"
)
```

The major differences between this and the compress function are that crf is set to be quality-based using a variable bitrate, fps mode is set by default to CFR, and pixel format and the moov atom is optimized for browser playback. So they do fulfill different functions.

ktshah04 replied:
I thin it would be helpful to break out the functionality of the compress and make_browser_friendly methods into a shared helper method and have both call that

Lucinda-Williamson committed:
Added a function to verify frame count computed by opencv.Useful when ingesting videos from external sources that may have corrupted metadata in the header

Lucinda-Williamson committed:
Added a varification to check opencv frame count against actual readable frames. Uses verify_frame_count to detect videos where container metadata reports more frames than can actually be read. Falls back to confirmed frame count so timestamps and downstream tables don't error out.

ktshah04 commented on pose_pipeline/utils/video_format.py:
```diff
@@ -27,3 +29,81 @@ def insert_local_video(filename, video_start_time, local_path, video_project="TE
+            first_bad = test_frame - 1
+
+    # last_good is a 0-indexed frame number, so actual count is last_good + 1
+    return last_good + 1
```
This would be good to fix

ktshah04 commented on pose_pipeline/utils/video_format.py:
```diff
@@ -27,3 +29,81 @@ def insert_local_video(filename, video_start_time, local_path, video_project="TE
+    ]
+
+    print("Running:", " ".join(cmd))
+    subprocess.run(cmd, check=True)
```
This would be good to fix

ktshah04 commented on pose_pipeline/pipeline.py:
```diff
@@ -113,6 +114,15 @@ def make(self, key, override=False):
+        confirmed_frames = verify_frame_count(cap, frames)
+        if confirmed_frames != frames:
+            import warnings
+            warnings.warn(
```
I think it could be worth considering raising an error vs just having a warning here. If we think this will be a rare issue then probably better to raise and error, otherwise warning is fine

Lucinda-Williamson replied:
I ended up sticking with the warning. I was testing and (1) if batching a bunch of keys it could be annoying to the user if the process halts in the middle of a batch, (2) this function calculate the amount of readable frames so raising an error essentially throws that value away and the problem persists.

ktshah04 commented on pose_pipeline/pipeline.py:
```diff
@@ -113,6 +114,15 @@ def make(self, key, override=False):
+            warnings.warn(
+                f"Frame count mismatch for {key}: metadata says {frames}, "
+                f"but only {confirmed_frames} frames are readable. Using actual count.")   
+            frames = confirmed_frames
```
This short circuit is a good suggestion but probably better to just set key['num_frames'] once at the bottom

Lucinda-Williamson committed:
Adressing PR feddback for VideoInfo.make

- only setting key['num_frames' once
- fixed edge case if videos return 0 or 1 frames, accpeted suggested edit and added if frames > 0
- switched from warning to raising an error

Lucinda-Williamson committed:
Addressing PR feedback comments from copilot and kunal

- added helper function (_run_ffmpeg) to run ffmpeg. Now both the compress function and make_browser_friendly call this.
- in make_browser_friendly addressed the backup_dir issue by adding a validation step to make sure it exists. rewrites to a tmp file and only replaces original on sucess or restores on failure
- verify_frame_count addded a check if reported frames is 0

Lucinda-Williamson committed:
Removed excess print out on error message

- the fix might not always be to run it through that function so don't want to print that to the user

Lucinda-Williamson committed:
Fixed typo made doc string more concise

Lucinda-Williamson committed:
Adressed feedback on VideoInfo.make() and video_format
- resolved ciruclar import by moving Video import inside of insert_local_video()
- firmed up make_browser_friendly behavior: validate base_dir exists, name backp with '_original' suffix to guard against overwriting it, force .mp4 output container, use temp file replace strategy with backup restore if ffmpg fails
- verify_frame_count: made variable names clearer
- in VideoInfo make, wrapped frame check and insert logic into try/finally block to guarantee cap.release() and os.remove() run on all exit paths, also reverted back to warning rather than exception because the confirmed frames is already being calcualted and it seemed a waste to just throw that value away. Also in testing if running a bunch of keys it could be annoying if just a few keys have issues and keep halting the make method.

Lucinda-Williamson committed:
Adjusting from raising an exception to a warning. It seems counterintuitive to compute the confirmed frames and then just throw it away
